### PR TITLE
dfe-analytics configuration table & web request user_id pseudonymisation script

### PR DESCRIPTION
**Pull request by stevenleggdfe**
Currently we have a number of services with pseudonymised user IDs in database tables streamed to BigQuery, but where the same IDs are not pseudonymised in the ```user_id``` field in web request events streamed to BigQuery. This means that none of the security benefits of pseudonymisation are realised because the ```user_id```s are still stored in BigQuery in their raw form anyway!

This PR provides:
- A stored procedure to convert all raw ```user_id```s in the events table to pseudonymised ```user_id```s
- To support this, a table of dfe-analytics configurations over time created using the new ```initialise_analytics``` ```event_type``` introduced in dfe-analytics 1.9.0
- Fixes ```data_functions.eventDataCreateOrReplace()``` function to handle strings that include quotation marks correctly
- Updates test data to use ```candidates``` table instead of ```subjects``` to allow this to be tested more easily

Once merged we will need to send each team the following instructions if they wish to pseudonymise their ```user_id```s:
1. Ensure that the ```id``` field in a database table that matches some of the values in the ```user_id``` field in your events table in BigQuery is in both your ```analytics.yml``` and ```analytics_pii.yml``` files.
2. Set ```DfE::Analytics.config.pseudonymise_web_request_user_id = true``` in your application. See https://github.com/DFE-Digital/dfe-analytics#user-id-pseudonymisation.
3. Run your Dataform pipeline.
4. Invoke the stored procedure called ```pseudonymise_request_user_ids``` located in the same dataset as your events table, setting its ```user_table``` parameter to the name of the table you identified in step 1.
5. Identify any parts of your Dataform codebase which join raw user_ids to pseudonymised user_ids using something like ```JOIN ... ON pseudonymised_user_id = TO_HEX(SHA256(raw_user_id))``` and update them so they no longer use ```TO_HEX()``` and ```SHA256()``` GoogleSQL functions to pseudonymise the raw ID before joining. This is necessary because the raw ID will no longer be stored in BigQuery.
6. Full refresh your Dataform pipeline.

### CODE REVIEW

**Review by alexchristensenedu — COMMENTED** [review]
Minor comment

**Comment by asatwal** [comment]
> 2\. DfE::Analytics.config.pseudonymise_web_request_user_id = false

In the PR description above, this is actually wrong. It refers to the def-analytics GEM which is also wrong I believe. I think the intention here is to start psuedonymising right ?


It should be:

`DfE::Analytics.config.pseudonymise_web_request_user_id = false`

File: includes/pseudonymise_request_user_ids.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  THEN RAISE USING MESSAGE = "Data from user_table does not appear to have been streamed by dfe-analytics.";
+END IF;
+
+IF
+  (
+  /* Detect whether entity_id appears to be SHA256-formatted (i.e. containing only digits and lower case letters, and at least one of each) */
+  SELECT
+    LOGICAL_AND((REGEXP_CONTAINS(entity_id, "[0-9]")
+        AND REGEXP_CONTAINS(entity_id, "[a-z]")
+        AND REGEXP_CONTAINS(entity_id, "^[0-9a-z]*$")))
```
[the remark this answers is not in the document]
Comment (stevenleggdfe) [reply includes/pseudonymise_request_user_ids.js]: Good spot, resolved!

**Comment by stevenleggdfe** [comment]
> > 2. DfE::Analytics.config.pseudonymise_web_request_user_id = false
> 
> In the PR description above, this is actually wrong. It refers to the def-analytics GEM which is also wrong I believe. I think the intention here is to start psuedonymising right ?
> 
> In which case it should be:
> 
> `DfE::Analytics.config.pseudonymise_web_request_user_id = true`

Good spot @asatwal - I've amended this. Will be important for this to be correct as I'll copy and paste these instructions into the release notes and instructions we send round teams.

**Review by alexchristensenedu — APPROVED** [review]

**Pull request merged**